processor/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+                Instruction::RET(Ret {})
+            }
+            _ => panic!("not match opcode:{}", opcode)
+        };
+        instuction
+    }
+
+    pub fn execute(&mut self, program: &mut Program, decode_flag: bool) -> Result<(), ProcessorError> {
+        let mut pc = 0;
+        loop {
+            let instruct_line = program.instructions[pc as usize].trim();
+            let mut txt_instruction = Default::default();
+            let mut step = 1;
+
+            if decode_flag {
+                let mut imm_line = "null";
+                if (pc + 1) < (program.instructions.len() as u64 - 1) {
+                    imm_line = program.instructions[(pc + 1) as usize].trim();
+                    (txt_instruction, step) = decode_raw_instruction(instruct_line, imm_line)?;
+                } else {
+                    (txt_instruction, step) = decode_raw_instruction(instruct_line, imm_line)?;
+                }
+                if step == IMM_INSTRUCTION_LEN {
+                    program.trace.raw_binary_instructions.push((instruct_line.to_string(), Some(imm_line.to_string())));
+                } else {
+                    program.trace.raw_binary_instructions.push((instruct_line.to_string(), None));
+                }
+
+                let instruction = self.decode_instruction(txt_instruction);
+                program.trace.raw_instructions.push(instruction);
+                pc += step;
+            }
+            if pc >= (program.instructions.len() as u64 - 1) {
+                break;
+            }
+        }
+        assert!(program.trace.raw_binary_instructions.len() == program.trace.raw_instructions.len());
+        loop {
+            let instruct_line = program.instructions[self.pc as usize].trim();
+
+            let mut instruction = Default::default();
+            let mut ops = Vec::new();
+            let mut step = 1;
+            if decode_flag {
+                let mut imm_line = "null";
+                if (self.pc + 1) < (program.instructions.len() as u64 - 1) {
+                    imm_line = program.instructions[(self.pc + 1) as usize].trim();
+                    (instruction, step) = decode_raw_instruction(instruct_line, imm_line)?;
+                } else {
+                    (instruction, step) = decode_raw_instruction(instruct_line, imm_line)?;
+                }
+                if step == IMM_INSTRUCTION_LEN {
+                    program.trace.raw_binary_instructions.push((instruct_line.to_string(), Some(imm_line.to_string())));
+                } else {
+                    program.trace.raw_binary_instructions.push((instruct_line.to_string(), None));
+                }
+            } else {
+                instruction = instruct_line.to_string();
+            }
+            ops = instruction.split(' ').collect();
+            let opcode = ops.get(0).unwrap().to_lowercase();
+
+            match opcode.as_str() {
+                "mov" => {
```

Comment:
Suggestion: It would be better to handle the instruction directly by enumeration, and not to enumerate the string.